### PR TITLE
Added class/method documentation.

### PR DESCRIPTION
**Pull request by fabianmarz**
I've added some class and method documentation as this was lacking and was not very understandable by others. Please review the docs and provide feedback. Thanks so much for you help! 🙏

### CODE REVIEW

File: src/Node/Render.php [hunk 1]
```diff
@@ -9,13 +13,31 @@
...
 
+  /**
+   * Compiles a template string.
+   *
+   * @param \Twig_Node_Expression $expr
+   * @param \Twig_Node_Expression|NULL $variables
+   * @param bool $only
+   * @param bool $ignoreMissing
+   * @param $lineno
+   * @param NULL $tag
```
Comment (sun) [inline src/Node/Render.php]: 1. After `@param` follows the data type, `NULL` is not a data type, but a value (or rather no value).

2. Unless this method signature was inherited from the parent class, each parameter should have a description.  If, on the other side, _all_ of the parameters are identical to the parent class, you may omit all of the `@param` tags.

File: src/Node/Render.php [hunk 2]
```diff
@@ -9,13 +13,31 @@
...
+ *
+ * Components are rendered within their own context. The default variables are
+ * set from the component config file if not given explicitly via the tag.
+ * Variants must be referenced with a double hyphen and be configured in the main
+ * component config. – Having a separate variant file is currently not supported.
+ */
 class Render extends Twig_Node_Include {
 
   /**
    * Holds the given component name from the render tag.
```
Comment (sun) [inline src/Node/Render.php]: For properties, we describe the value directly, without verbs like "holds" or the like.

```
The component name, as specified in the render tag.
```
However, if this is the component to render, then I'd simply state so:
```
The name of the component to render.
```

File: src/Node/Render.php [hunk 3]
```diff
@@ -9,13 +13,31 @@
...
    */
   protected $component;
 
+  /**
+   * Compiles a template string.
```
Comment (sun) [inline src/Node/Render.php]: A class constructor normally does not immediately perform any actions.  We simply state what is being constructed; for example:

```
Constructs a Twig template to render.
```

File: src/Node/Render.php [hunk 4]
```diff
@@ -9,13 +13,31 @@
...
+   * @param \Twig_Node_Expression $expr
+   * @param \Twig_Node_Expression|NULL $variables
+   * @param bool $only
+   * @param bool $ignoreMissing
+   * @param $lineno
+   * @param NULL $tag
+   */
   public function __construct(Twig_Node_Expression $expr, Twig_Node_Expression $variables = NULL, $only = FALSE, $ignoreMissing = FALSE, $lineno, $tag = NULL) {
     $this->component = $expr->getAttribute('value');
     // Remove variant appendix from component name as we do not have variant files.
```
Comment (sun) [inline src/Node/Render.php]: 1. "appendix" should be "suffix".

2. Shouldn't "component name" be "template name"…?

```diff
-Remove variant appendix from component name as we do not have variant files.
+Remove any variant suffixes from the template name as there are no template files for variants, only for components.
```

3. The following todo should be removed from code and converted into a GitHub issue instead.

4. If I understand it correctly, the following inline comment should be replaced with

```diff
-Add an 'not_used' node if we don't have any variables to render correctly.
+Variables cannot be empty, so assign a dummy placeholder variable.
```

5. Why are variables mandatory?  Is it because your custom `addTemplateArguments()` method in this class attempts to compile and `array_merge()` them unconditionally, without checking whether there are any first?

File: src/Node/Render.php [hunk 5]
```diff
@@ -44,9 +68,12 @@ protected function addTemplateArguments(Twig_Compiler $compiler) {
...
   /**
-   * Returns the component context from the component config file.
+   * Returns the component context variables from its config file.
    *
    * @todo Merge specific keys to prevent config duplication.
+   *
+   * @param array $component
+   *   The component context variables.
    */
   public function getComponentContext($component) {
```
Comment (sun) [inline src/Node/Render.php]: The function argument should be type-hinted with `array`.

File: src/Node/Render.php [hunk 6]
```diff
@@ -44,9 +68,12 @@ protected function addTemplateArguments(Twig_Compiler $compiler) {
   }
 
   /**
-   * Returns the component context from the component config file.
+   * Returns the component context variables from its config file.
    *
    * @todo Merge specific keys to prevent config duplication.
```
Comment (sun) [inline src/Node/Render.php]: What's meant with this todo?

Note that most todos should either be follow-up tickets in Asana (if time-critical) or GitHub (if not).

File: src/Node/Render.php [hunk 7]
```diff
@@ -29,7 +51,9 @@ public function __construct(Twig_Node_Expression $expr, Twig_Node_Expression $va
   }
 
   /**
-   * Adds the template arguments including the component (variant) context.
+   * Adds the component context variables to the Twig compiler string.
```
Comment (sun) [inline src/Node/Render.php]:
```
Adds the template variables to the compiled Twig PHP template string.

Variables consist of

1. the default `context` properties in the component's config file,
   which may be overridden by
2. the `context` properties of the requested variants (if any) in the
   component's config file, which may be overridden by
3. the variables passed to the `render` tag itself.
```

File: src/Node/Render.php [hunk 8]
```diff
@@ -100,7 +144,13 @@ public function getVariantContext($variant, $config) {
...
+   * Returns the sanitized component name without its variant.
+   *
+   * @param string $component
+   *   The component name to sanitize.
+   *
+   * @return string
+   *   The base component name without its variant.
    */
   public function sanitizeComponentName($component) {
     return preg_replace('@--[a-z0-9-]+@', '', $component);
```
Comment (sun) [inline src/Node/Render.php]: We probably want to ignore letter-casing, so we should add the modifier `i`.

File: src/Node/Render.php [hunk 9]
```diff
@@ -100,7 +144,13 @@ public function getVariantContext($variant, $config) {
...
-   * Returns the sanitized component name
+   * Returns the sanitized component name without its variant.
+   *
+   * @param string $component
+   *   The component name to sanitize.
+   *
+   * @return string
+   *   The base component name without its variant.
    */
   public function sanitizeComponentName($component) {
```
Comment (sun) [inline src/Node/Render.php]: Architecturally, it does not make sense to execute this method more than once.  Also, "sanitize" is a misleading name, the term is only used for cleaning/escaping operations.  The operation performed here is much rather "extract".

Can we determine this information directly in the constructor and set the result to properties?

```php
$this->variants = explode('--', $component);
$this->component = array_shift($this->variants);
```

That's essentially what `getComponentParts()` is doing, but supports multiple variants.  If you prefer to keep a separate method for unit testing purposes, then I suggest to make it
```
return [$component, $variants];
```
and perform the assignment to class properties in the calling code (constructor).

File: src/Node/Render.php [hunk 10]
```diff
@@ -85,6 +120,15 @@ public function getComponentParts($component) {
...
    * Returns the variant context from the config file.
+   *
+   * @param string $variant
+   *   The variant name to get the context for.
+   * @param string $config
+   *   The component config where to lookup the variant in.
+   *
+   * @return array
+   *   The variant context variables.
+   *
```
Comment (sun) [inline src/Node/Render.php]: Superfluous blank line at the end.

File: src/Node/Render.php [hunk 11]
```diff
@@ -9,13 +13,31 @@
...
 use Twig_Node_Include;
 
+/**
+ * Class to create and compile a new 'Render' node Twig template.
+ *
+ * Components are rendered within their own context. The default variables are
+ * set from the component config file if not given explicitly via the tag.
+ * Variants must be referenced with a double hyphen and be configured in the main
+ * component config. – Having a separate variant file is currently not supported.
+ */
```
Comment (juliquiron) [inline src/Node/Render.php]: In a first reading of it, my dobuts:
 * What is a context?
 * Which config file are you talking about?
 * What variant is?
Maybe we should refer here the whole `fractal variants` thing to explain what are we adding support for. Also a `@see` with it will make sense.

File: src/Node/Render.php [hunk 12]
```diff
@@ -44,9 +68,12 @@ protected function addTemplateArguments(Twig_Compiler $compiler) {
   }
 
   /**
-   * Returns the component context from the component config file.
+   * Returns the component context variables from its config file.
    *
    * @todo Merge specific keys to prevent config duplication.
+   *
```
Comment (juliquiron) [inline src/Node/Render.php]: why we are not achieving this `@todo` right now?

File: src/Node/Render.php [hunk 13]
```diff
@@ -44,9 +68,12 @@ protected function addTemplateArguments(Twig_Compiler $compiler) {
...
+   * @param array $component
+   *   The component context variables.
```
Comment (juliquiron) [inline src/Node/Render.php]: Is a `@return` missing here?

File: src/Node/Render.php [hunk 14]
```diff
@@ -64,6 +91,8 @@ public function getComponentContext($component) {
 
   /**
    * Returns the component config path.
+   *
+   * @param $component
```
Comment (juliquiron) [inline src/Node/Render.php]: Is a `@return` missing here?

File: src/TokenParser/Render.php [hunk 15]
```diff
@@ -1,19 +1,50 @@
...
+ */
 
 namespace Drupal\twig_fractal\TokenParser;
 
 use Drupal\twig_fractal\Node;
 use Twig_Token;
 use Twig_TokenParser;
 
+/**
+ * Registers a 'render' tag and returns a new 'Render' node when the tag is parsed.
```
Comment (juliquiron) [inline src/TokenParser/Render.php]: Is the class registring it or implementing it? Is the class who returns the 'Render' or is a method. Maybe some more functional side description will be better for a class level doc?

File: src/TokenParser/Render.php [hunk 16]
```diff
@@ -25,11 +56,15 @@ protected function parseArguments() {
     if ($stream->nextIf(Twig_Token::NAME_TYPE, 'with')) {
       $variables = $this->parser->getExpressionParser()->parseExpression();
     }
+    // Strictly limit to own component context.
```
Comment (juliquiron) [inline src/TokenParser/Render.php]: This comment to which lines affect? The 3? or just the 2nd ones?

File: src/TwigFractal.php [hunk 17]
```diff
@@ -1,20 +1,48 @@
...
+ *
+ * Example usage:
+ * @code
+ * {% render '@foo/baz.twig' %}
+ * {% render '@foo/baz.twig' with { qux: 'foobar' } %}
+ * {% render '@foo/baz--bar.twig' with { qux: 'foobar' } %}
+ * @endcode
+ *
+ * @implements Twig_ExtensionInterface
+ */
```
Comment (juliquiron) [inline src/TwigFractal.php]: Maybe here should we also refer the render class we already implemented? IMO some class should have to the fractal variants thing project/spects/api/... not sure about which one makes more sense to have it.

**Review by juliquiron — CHANGES_REQUESTED** [review]
Great work! I added some considerations to the comments where IMO can be improved or more useful.

File: src/Node/Render.php [hunk 18]
```diff
@@ -85,6 +120,15 @@ public function getComponentParts($component) {
 
   /**
    * Returns the variant context from the config file.
+   *
+   * @param string $variant
+   *   The variant name to get the context for.
+   * @param string $config
+   *   The component config where to lookup the variant in.
```
Comment (sun) [inline src/Node/Render.php]: Wouldn't it be more appropriate to pass just the array of variants instead of the whole config?

File: src/Node/Render.php [hunk 19]
```diff
@@ -73,6 +102,12 @@ protected function getComponentConfig($component) {
 
   /**
    * Splits the component and return its parts.
+   *
+   * @param $component string
```
Comment (sun) [inline src/Node/Render.php]: The data type comes first, not second.

File: src/Node/Render.php (hunk 14 above)
Comment (sun) [inline src/Node/Render.php]: Missing data type.

File: src/Node/Render.php [hunk 20]
```diff
@@ -64,6 +91,8 @@ public function getComponentContext($component) {
 
   /**
    * Returns the component config path.
```
Comment (sun) [inline src/Node/Render.php]:
```diff
-Returns the component config path.
+Returns the file path of the Fractal YAML configuration file for a given component name.
```

Comment (sun) [reply src/Node/Render.php]: Is the file path absolute or relative to the document root?

File: src/Node/Render.php (hunk 5 above)
Comment (sun) [reply src/Node/Render.php]: In the code of this function, you can use the `+=` operator to append non-existing keys to an array.

```diff
-$context = $context + $this->getVariantContext($component['variant'], $config);
+$context += $this->getVariantContext($component['variant'], $config);
```

File: src/Node/Render.php [hunk 21]
```diff
@@ -9,13 +13,31 @@
 use Twig_Node_Expression_Constant;
 use Twig_Node_Include;
 
+/**
+ * Class to create and compile a new 'Render' node Twig template.
+ *
+ * Components are rendered within their own context. The default variables are
+ * set from the component config file if not given explicitly via the tag.
+ * Variants must be referenced with a double hyphen and be configured in the main
+ * component config. – Having a separate variant file is currently not supported.
```
Comment (sun) [inline src/Node/Render.php]: This is a good start, but I'd suggest to rephrase the summary line and description with the following in mind:

1. For class-level docs, don't assume that the reader knows anything about internals and how things work and how they relate to each other; all of that should be explained.

1. In general, it's hard to follow the current docs, because a "context" in Fractal means "default variables", whereas Drupal calls that "variables" (template variables).  A "context" in Drupal usually refers to the surrounding scope and environment of the running script; i.e., the execution context.  I'd recommend to always refer to "Fractal context properties" to make that distinction more clear, or simplify "default variables from component definition".

1. I'm purposively trying to avoid the term "configuration" as well, because that is universally associated with the actual site configuration in Drupal.  It is not recommended to refer to configuration unless you actually mean configuration.  Fractal's config files are much rather _component declaration files_ or _component definition files_.

1. I'm relatively sure that there is some terminology from Twig internals that additionally confuses matters even more — such as "node", which probably means "tag" here?

1. How does a Render node differ from an Include node?  It's worth to briefly state the history, why this separate render tag was created.  From a bird-level perspective, that might actually be the best approach for class-level docs.

**Review by sun — CHANGES_REQUESTED** [review]

File: src/Node/Render.php (hunk 11 above)
Comment (sun) [reply src/Node/Render.php]: Good to see that we're raising more or less the same questions – I didn't read your review comments before my review in https://github.com/netzstrategen/drupal-twig_fractal/pull/1#discussion_r134539185 👍

File: src/TwigFractal.php (hunk 17 above)
Comment (sun) [reply src/TwigFractal.php]: @juliquiron Can you clarify what you mean here?

File: src/Node/Render.php (hunk 7 above)
Comment (sun) [reply src/Node/Render.php]: The suggested PHPDoc makes sense for explaining the intended internals of the method here.

In addition, we should also consider to copy an explanation like this into the bird-level documentation, which currently seems to live on the main TwigFractal class.

File: src/TwigFractal.php (hunk 17 above)
Comment (juliquiron) [reply src/TwigFractal.php]: I mean that this class is basically a wrapper to introduce the functionality of the `Drupal\twig_fractal\TokenParser\Render` so maybe in the class level description makes sense to some reference to it. Anyway I'm not sure if this make sense, so just sharing my thinks as a 'developer how just read this file'

File: src/Node/Render.php (hunk 13 above)
Comment (fabianmarz) [reply src/Node/Render.php]: This was completely wrong. Added several fixup commits where I clarified this.

File: src/Node/Render.php (hunk 12 above)
Comment (fabianmarz) [reply src/Node/Render.php]: Created an issue https://github.com/netzstrategen/drupal-twig_fractal/issues/3.

File: src/Node/Render.php (hunk 4 above)
Comment (fabianmarz) [reply src/Node/Render.php]: You're right! Thanks for catching this. I several fixup commits for this.

**Comment by fabianmarz** [comment]
Really appreciate your feedback @sun and @juliquiron. I added several fixup commits to clarify the code.

**Comment by sun** [comment]
@fabianmarz Added a commit to clean up various details.  This looks ready to me now.  Thanks for your patience!  I'd suggest to squash this PR into two commits and proceed with merging into master.

**Review by sun — APPROVED** [review]
